vterebey/budgetcalculator/budget_calculator_class.py

```diff
@@ -0,0 +1,302 @@
+"""
+'Budget Calculator' program to track expenses by category and date.
+To use you need to run the program. Next you need to choose what you
+are going to do, you can:
+1. Enter or select an expense category and cost
+2. View spending statistics on a pie chart
+3. Display expenses by category
+(you can choose the suggested one or enter it yourself)
+4. Display expenses by month (you must enter the year and
+month in the format: YYYY-MM)
+5. Display expenses by category and month at the same time
+(see above how the category and month are set)
+All expenses are recorded in the file 'values.csv'
+The file contains the category,date and cost
+"""
+
+
+import csv
+import re
+from datetime import datetime
+import matplotlib.pyplot as plt
+
+
+# regular expressions for category validation
+fieldnames = ["Category", "Date", "Amount"]
+CAT_REGEX = "^[0-9]+$"
+cat_pattern = re.compile(CAT_REGEX)
+default_categories = {
+    1: "Grocery store",
+    2: "Clothing",
+    3: "Entertainment",
+    4: "Cafe",
+    5: "Store",
+    6: "Pharmacy",
+}
+with open("values.csv", "w", encoding="UTF-8", newline="") as file:
+    writer = csv.writer(file)
+    writer.writerow(fieldnames)
+
+
+class Budget:
+
+    "A class to represent a budget"
+
+    categories = {}
+
+    def __init__(self, category, date, amount):
+        """
+        This function initializes expense values
+        """
+        self.category = category
+        self.amount = amount
+        self.date = date
+
+    def _add_expense(self):
+        """
+        This function adds category and it's amount to the dictionary
+        for displaying statistics
+        """
+        if self.category not in Budget.categories:
+            Budget.categories[self.category] = self.amount
+        else:
+            Budget.categories[self.category] += self.amount
+
+    def expense_record(self):
+        """This function records expenses to a file"""
+        if int(self.amount) == self.amount:
+            self.amount = int(self.amount)
+        with open("values.csv", "a", encoding="UTF-8", newline="") as csvfile:
+            writer_ = csv.writer(csvfile)
+            writer_.writerow([self.category, self.date, self.amount])
+        Budget._add_expense(self)
+
+    def expenses_stat(self):
+        """This function dispays expenses in the form of a chart"""
+        amounts = list(Budget.categories.values())
+        # pylint: disable=W0612
+        fig1, ax1 = plt.subplots()
+        ax1.pie(
+            amounts,
+            labels=list(Budget.categories.keys()),
+            autopct="%1.1f%%",
+            shadow=False,
+            startangle=90,
+        )
+        ax1.axis("equal")
+        ax1.set_title("Statistics")
+        plt.show()
+
+    def group_by_cat(self, category):
+        """This function groups expenses by category"""
+        with open("values.csv", "r", encoding="UTF-8", newline="") as csvfile:
```

Comment:
it might be better to open file only once, in the beggining, and re-read it only if we updated data inside

Reply:
Updated